### DISCUSSION TITLE
How does node.js code interact with libUV?

### DISCUSSION THREAD

**Discussion by dSalieri** [libuv/libuv#3784] [Q&A]
I've tried to find something about that how node.js interact with libUV. But I didn't find anything good as explanation.

Well, if we take for example node.js' stream and destroy it intentionally by `stream.destroy()` (`destroy` is a method of node.js library that belong to streams). So that method must destroy our stream, but it doesn't, take a look at the [source of destroy's code](https://github.com/nodejs/node/blob/2b32985c626b57d7ea3e68d6ce088ea3183e3a58/lib/internal/streams/destroy.js). There is no something that could be destroy our stream. So how does node.js interconnect with libUV

P.S I know that it is about node architecture, but who knows maybe somebody knows something about my question

**Comment by bnoordhuis** [comment]
Streams in node are generalized. There's the public API that you interact with as a consumer of a stream, and then there are APIs for implementers of streams.

Calling `stream.destroy()` on a stream that's backed by a TCP or UNIX socket (named pipe on Windows) does eventually end up calling `uv_close()` but there are quite a few layers in between. Run `git grep uv_ src` on node's source tree to find examples.

**Reply by dSalieri** [reply]
@bnoordhuis this is very interesting... but I am trying to find out how the node's implementation of streams connecting with libUV. Where is that bridge between `stream.js` and [`stream.c`](https://github.com/nodejs/node/blob/ceb8f9c6eb9aeb515f112e9ba2115057b6ccea62/deps/uv/src/unix/stream.c), particulary between [`destroy.js`](https://github.com/nodejs/node/blob/v18.10.0/lib/internal/streams/destroy.js) and `uv_close()`

I am interesting it because I want to know in what gap of the time node.js solve to destroy my ordinary stream. I see invoking close's event on the stream [`self.emit("close")`](https://github.com/nodejs/node/blob/v18.10.0/lib/internal/streams/destroy.js#L132) but don't see destroying of it... that moment confusing me.

I need help.

**Reply by bnoordhuis** [reply]
Node's libuv integration is kind of all over the place. There's no single entry point I can point you to but you may want to start by reading src/handle_wrap.cc and src/stream_wrap.cc.

**Comment by theanarkh** [comment]
For example. 
1. TCP socket has a field `_handle`, which is a object related to a C++ `TCPWrap` object.
2. the `TCPWrap` object has a field `handle_` which is a `uv_tcp_t`. `uv_tcp_t` is a `Libuv` data structure.
3. when we destroy JS socket, it will call `socket._handle.close`, `close` function is defined in `handle_wrap.cc`, and the value is `HandleWrap::Close`, then `HandleWrap::Close` call `uv_close(handle_, ...)`.

I hope this can help you.
